File: yarl-core-java/src/test/java/hr/tjakopan/yarl/retry/WaitAndRetryForeverHandleResultTest.java
package hr.tjakopan.yarl.retry;

import hr.tjakopan.yarl.test.helpers.TestResult;
import org.junit.Test;

import java.time.Duration;
import java.util.ArrayList;
import java.util.HashMap;

import static hr.tjakopan.yarl.Functions.fromConsumer4;
import static org.assertj.core.api.Assertions.assertThat;

public class WaitAndRetryForeverHandleResultTest {
  @Test
  public void shouldBeAbleToCalculateRetryDurationsBasedOnTheHandledFault() {
    final var expectedRetryWaits = new HashMap<TestResult, Duration>() {{
      put(TestResult.FAULT, Duration.ofMillis(2));
      put(TestResult.FAULT_AGAIN, Duration.ofMillis(4));
    }};
    final var actualRetryWaits = new ArrayList<Duration>(2);
    final var policy = RetryPolicy.<TestResult>builder()
      .handleResult(TestResult.FAULT)
      .handleResult(TestResult.FAULT_AGAIN)
      .waitAndRetryForever((i, outcome, c) -> outcome.fold(expectedRetryWaits::get, e -> Duration.ZERO),
        fromConsumer4(d -> duration -> (i, c) -> actualRetryWaits.add(duration)));

    final var iterator = expectedRetryWaits.keySet().iterator();
    policy.execute(() -> {
      if (iterator.hasNext()) {
        return iterator.next();
      } else {
        return TestResult.UNDEFINED;
      }
    });

    assertThat(actualRetryWaits).containsExactlyElementsOf(expectedRetryWaits.values());
  }
}

File: yarl-core/src/test/kotlin/hr/tjakopan/yarl/retry/WaitAndRetryForeverHandleResultTest.kt
package hr.tjakopan.yarl.retry

import hr.tjakopan.yarl.Policy
import hr.tjakopan.yarl.test.helpers.TestResult
import org.assertj.core.api.Assertions.assertThat
import java.time.Duration
import kotlin.test.Test

class WaitAndRetryForeverHandleResultTest {
  @Test
  fun shouldBeAbleToCalculateRetryDurationsBasedOnTheHandledFault() {
    val expectedRetryWaits =
      mapOf(TestResult.FAULT to Duration.ofMillis(2), TestResult.FAULT_AGAIN to Duration.ofMillis(4))
    val actualRetryWaits = mutableListOf<Duration>()
    val policy = Policy.retry<TestResult>()
      .handleResult(TestResult.FAULT)
      .handleResult(TestResult.FAULT_AGAIN)
      .waitAndRetryForever({ _, outcome, _ ->
        outcome.fold({ expectedRetryWaits[it] }, { Duration.ZERO }) ?: Duration.ZERO
      }) { _, duration, _, _ -> actualRetryWaits.add(duration) }

    val iterator = expectedRetryWaits.iterator()
    policy.execute {
      if (iterator.hasNext()) {
        return@execute iterator.next().key
      } else {
        return@execute TestResult.UNDEFINED
      }
    }

    assertThat(actualRetryWaits).containsExactlyElementsOf(expectedRetryWaits.values)
  }
}
